examples/Coffee_Machine_FSM/coffee_active_learning.py

```diff
@@ -0,0 +1,184 @@
+logger = logging.getLogger(__name__)
+
+# Definition of Action and Observation Types
+Action = str  # Possible actions from coffee_fsm: 'pod', 'water', 'button', 'clean'
+
+@dataclass
+class CoffeeObservation:
+    status_code: StatusCode
+    previous_state: str
+    current_state: str
+
+# ============================
+# CoffeeEnvironment Class
+# ============================
+class CoffeeEnvironment(ActiveEnvironment[Action, CoffeeObservation]):
+    def __init__(self, max_num_iterations: int):
+        self.coffee_machine = CoffeeMachine()
+        self.last_action: Optional[Action] = None
+        self.status_code = StatusCode.OK
+        self.max_num_iterations = max_num_iterations
+        self.current_iteration = 0
+        self.previous_state = self.coffee_machine.get_current_state()
+
+    def load(self) -> 'CoffeeEnvironment':
+        return self
+
+    def act(self, action: Action) -> None:
+        self.last_action = action
+
+    def step(self) -> None:
+        if self.last_action is None:
+            raise ValueError("No action taken yet")
+        # Record the state before the action
+        self.previous_state = self.coffee_machine.get_current_state()
+        output_sequence = self.coffee_machine.transition(self.last_action)
+        self.status_code = StatusCode(output_sequence)
+        logger.debug(f"Action: {self.last_action}, Previous State: {self.previous_state}, Status: {self.status_code}")
+        self.current_iteration += 1
+        if self.current_iteration >= self.max_num_iterations:
+            raise StopLearning()
+        if self.status_code == StatusCode.ERROR or self.status_code == StatusCode.COFFEE_PRODUCED:
+            # Reseting the coffee machine to initial state after error or coffee produced
+            self.coffee_machine.current_state = StateA()
+
+    def observe(self) -> CoffeeObservation:
+        return CoffeeObservation(
+            status_code=self.status_code,
+            previous_state=self.previous_state,
+            current_state=self.coffee_machine.get_current_state(),
+        )
+
+# ============================
+# CoffeeModel Class
+# ============================
+class CoffeeModel(Model):
+    def __init__(self):
+        # Storing the best action per state
+        self.best_actions: Dict[str, Action] = {}
+
+    def predict(self, input_features: pl.DataFrame) -> pl.DataFrame:
+        actions = []
+        for state in input_features['current_state']:
+            if state in self.best_actions:
+                actions.append(self.best_actions[state])
+            else:
+                # Default action if state is unknown
+                actions.append('clean')
+        return pl.DataFrame({'action': actions})
+
+    def save(self, path: Path) -> None:
+        pass
+
+    def load(self, path: Path) -> None:
+        pass
+
+# ============================
+# CoffeeLearner Class
+# ============================
+class CoffeeLearner(ActiveLearner[Action, CoffeeObservation]):
```

Comment:
Again, this could be part of the learner-directory, which is maybe not necessary, here.

The class can be more general, again.